client/data/home/use-skip-current-view-mutation.ts

```diff
@@ -0,0 +1,66 @@
+/**
+ * External dependencies
+ */
+import { useCallback } from 'react';
+import { useMutation, useQueryClient, UseMutationResult } from 'react-query';
+
+/**
+ * Internal dependencies
+ */
+import wp from 'calypso/lib/wp';
+import { getHomeLayoutFlags } from './home-layout-flags';
+
+type ReminderDuration = '1d' | '1w' | null;
+
+interface Variables {
+	siteId: number;
+	viewName: string;
+	reminder: ReminderDuration;
+}
+
+interface Result extends UseMutationResult< void, unknown, Variables > {
+	skipCurrentView: ( siteId: number, reminder: ReminderDuration ) => void;
+}
+
+function useSkipCurrentViewMutation(): Result {
+	const queryClient = useQueryClient();
+	const { isDev } = getHomeLayoutFlags();
+
+	const mutation = useMutation< void, unknown, Variables >(
+		( { siteId, viewName, reminder } ) =>
+			wp.req.post(
+				{
+					path: `/sites/${ siteId }/home/layout/skip`,
+					apiNamespace: 'wpcom/v2',
+				},
+				{ ...( isDev && { dev: true } ) },
+				{
+					view: viewName,
+					...( reminder && { reminder } ),
+				}
+			),
+		{
+			onSuccess( data, { siteId } ) {
+				queryClient.setQueryData( [ 'home-layout', siteId ], data );
+			},
+		}
+	);
+
+	const { mutate } = mutation;
+
+	const skipCurrentView = useCallback(
+		( siteId: number, reminder: ReminderDuration ) => {
+			const data = queryClient.getQueryData( [ 'home-layout', siteId ] );
```

Comment:
This code relies on someone else to fire the `useHomeLayoutQuery` query and save the response to cache. If nobody fires the query or if `skipCurrentView` is called before the query response arrives, the mutation will silently do nothing.

Code like this would be more reliable:
```js
const layout = await wpcom.req.get( '/sites/${ id }/home/layout' );
await wpcom.req.post( '/sites/${ id }/home/layout/skip', { view: layout.view_name } );
```
Only we don't want to use `wpcom.req` directly, but wrapped in React Query so that cached data are used if available.

This code makes sure that the `layout` is really fetched, waits if it needs to wait, and always performs the mutation with the right `view` parameter.

Reply:
I was able to make this mutation more reliable by fetching the current view in the mutation function:
```js
const mutation = useMutation(
  async ( { reminder } ) => {
    const data = await queryClient.fetchQuery(
      [ 'home-layout', siteId ],
      () => fetchHomeLayout( siteId, query ),
      { staleTime: Infinity }
    );
    const view = data.view_name;
    return await wp.req.post( `/sites/${ siteId }/skip`, { view, reminder } );
  },
  {
    onSuccess( data ) {
      queryClient.setQueryData( ... );
    }
  }
);
```

The `fetchHomeLayout` function is extracted and exported from `useHomeLayoutQuery`, being reused in both hooks.

The `queryClient.fetchQuery` will return data from cache, if it was already fetched and stored, or will issue the fetch request itself.

The `staleTime: Infinity` option makes sure that `fetchQuery` never re-fetches if the data are already in cache. The default `staleTime` is 0, causing a refetch on every call.

Reply:
Nice! I was thinking of calling `useHomeLayoutQuery()` from within the body of `useSkipCurrentViewMutation()`, but I didn't want to pre-emptively fetch until someone had actually called the mutate function. This gets around this nicely. I very much like that ReactQuery has a non-React API like this.